fuel-indexer/api_server/src/main.rs

```diff
@@ -0,0 +1,81 @@
+cfg_if::cfg_if! {
```

Comment:
What would happen if someone tried to build this project with `--all-features` enabled? This style of feature flagging seems potentially flakey. I think the database type should be selectable at runtime, rather than rely on compilation options. This way we could package a single docker image that could run in either configuration later on for example.

Reply:
I guess I could copy/paste models.rs and make two identical versions of it, but that seems like an even uglier solution. The main issue there is the `use` statements import different schema.rs files for the different database backends, and all that logic should be identical between the two. There's nothing stopping us from just building it twice and then wrapping a simple shell script around the two.

Reply:
Even if it was uglier, I think it'd be preferable in order to avoid the footguns associated of not being able to do things like `cargo test --all-features` where you could potentially have large swaths of the code go untested (which has happened before in fuel crates and is why we are taking a hard stance on this now of always being able to use `--all-features`).

Fortunately, with some quick poking around of other projects depending on diesel it appears that hyperledger sawtooth already solved a similar problem with clever use of traits and connection wrappers.

They were able to unify most of their schema and make it SQL agnostic: https://github.com/hyperledger/transact/blob/bfe7b3fc73d9fc2b6f32a95cf63a4f047ec1fe2a/libtransact/src/state/merkle/sql/store/schema.rs#L18

In the areas where it was SQL dialect specific, rather than relying on the models directly, they use traits which specify the kinds of operations they'd like to do, and then implement them for each type of possible SQL connection. Here's an example of where they have two "conflicting" tables:

[postgres_merkle_radix_tree_node](https://github.com/hyperledger/transact/blob/bfe7b3fc73d9fc2b6f32a95cf63a4f047ec1fe2a/libtransact/src/state/merkle/sql/store/schema.rs#L48)
[sqlite_merkle_radix_tree_node](https://github.com/hyperledger/transact/blob/bfe7b3fc73d9fc2b6f32a95cf63a4f047ec1fe2a/libtransact/src/state/merkle/sql/store/schema.rs#L36)

And here is where they use alternate implementations of the conflicting models by implementing the same trait on different connection types:

[sqlite connection impl](https://github.com/hyperledger/transact/blob/e2def80b9c57f505bc8aaa0cbf98745b87d283d0/libtransact/src/state/merkle/sql/store/operations/write_changes.rs#L62)
[pg connection impl](https://github.com/hyperledger/transact/blob/e2def80b9c57f505bc8aaa0cbf98745b87d283d0/libtransact/src/state/merkle/sql/store/operations/write_changes.rs#L188)